pest/src/iterators/pairs.rs

```diff
@@ -255,7 +255,7 @@ impl<'i, R: Eq> Eq for Pairs<'i, R> {}
 impl<'i, R: Hash> Hash for Pairs<'i, R> {
     fn hash<H: Hasher>(&self, state: &mut H) {
         (&*self.queue as *const Vec<QueueableToken<R>>).hash(state);
-        (self.input as *const [u8]).hash(state);
+        (self.input.as_bytes() as *const [u8]).hash(state);
```

Comment:
(Disclaimer: implementation is a Rust implementation detail)

```rust
    #[stable(feature = "rust1", since = "1.0.0")]
    impl<T: ?Sized> Hash for *const T {
        fn hash<H: Hasher>(&self, state: &mut H) {
            if mem::size_of::<Self>() == mem::size_of::<usize>() {
                // Thin pointer
                state.write_usize(*self as *const () as usize);
            } else {
                // Fat pointer
                let (a, b) = unsafe {
                    *(self as *const Self as *const (usize, usize))
                };
                state.write_usize(a);
                state.write_usize(b);
            }
        }
    }
```

The observed effect is that [either approach will hash both address and length](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=a0eb5d197b361908e36b1243b1c67091). If you just want to hash the address, use str::as_ptr.

Which is the better approach here? I agree that if the same behavior is desired, `as *const str` is better than `.as_bytes() as *const [u8]`, but this is a good time to clarify which is desired.